Avoid using __TIMESTAMP__ when building for a compilation cache

emw-apple opened the pull request:
#### adb96155be5e75707c6d831c3329b72b0afcc2b1
<pre>
Avoid using __TIMESTAMP__ when building for a compilation cache
<a href="https://bugs.webkit.org/show_bug.cgi?id=247622">https://bugs.webkit.org/show_bug.cgi?id=247622</a>
rdar://102056857

Reviewed by Mark Lam.

For Xcode-based ports, add a build rule to generate a
&quot;JSCBytecodeCacheVersion.cpp&quot; source file, which contains a SHA1 hash of
CachedTypes.o and JSCBuiltins.o, plus a production source version string
if specified. This is a belt-and-suspenders approach to try and ensure
that the version hash changes any time the JSC sources change, while
still being reproducible.

Because of its input and output dependencies, XCBuild runs the build
rule after CachedTypes.cpp and JSCBuiltins.cpp are compiled, then
subsequently compiles the source file the build rule generates.

For CMake + Ninja, it doesn&apos;t seem to be easily possible to do achieve
the same build ordering without splitting CachedTypes and JSCBuiltins
out to a separate target. For now, continue to use __TIMESTAMP__.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj: Add
  build rule and source references.
* Source/JavaScriptCore/Sources.txt: Remove CachedTypes.cpp and
  JSCBuiltins.cpp from the unified source set, so that they get their
  own object files.
* Source/JavaScriptCore/runtime/CachedTypes.cpp:
(JSC::GenericCacheEntry::isUpToDate const):
(JSC::jscBytecodeCacheVersion): Deleted.
* Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.cpp.in: Added.
  Template file, contains a CACHED_TYPES_CKSUM token which is replaces
  with the actual checksum by the build rule.
* Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.h: Added.
* Source/ThirdParty/libwebrtc/Source/third_party/yasm/frontends/yasm/yasm.c:
  Remove __DATE__ macro used in assembler output.

Canonical link: <a href="https://commits.webkit.org/256816@main">https://commits.webkit.org/256816@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b84231e6534bb669de055f87bfafcc1eba8d0ff1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96256 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5511 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29306 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105802 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166149 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5632 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34264 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88633 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/102540 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101927 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/4172 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82852 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31196 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86015 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87931 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74020 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/87254 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39988 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/19449 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/82635 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37665 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20808 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28277 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4723 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/84 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/43403 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/85315 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/92 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40075 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19250 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
```diff
@@ -2173,6 +2176,20 @@
+			outputFiles = (
+				"$(DERIVED_FILE_DIR)/JSCBytecodeCacheVersion.cpp",
+			);
+			script = "CKSUM=($(cksum \"${SCRIPT_INPUT_FILE_0}\")) &&\nsed -e s/CACHED_TYPES_CKSUM/${CKSUM[0]}/ \"${INPUT_FILE_PATH}\" > \"${SCRIPT_OUTPUT_FILE_0}\"\n";
```
You probably need to do something to avoid breaking CMake?

aproskuryakov commented on Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.h:
```diff
@@ -0,0 +1,26 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+extern unsigned JSCBytecodeCacheVersion;
```
Is the version written with a size, or is this field's size effectively immutable? Not clear if it can be easily extended to be longer.

emw-apple replied:
It's compared against a `uint32_t m_cacheVersion`, so perhaps I should just use a uint32 here too. I was copying from the size type used by `WTF::StringHasher`.

MenloDorian commented on Source/JavaScriptCore/Sources.txt:
```diff
@@ -770,7 +770,6 @@ runtime/CacheUpdate.cpp
 runtime/CacheableIdentifier.cpp
 runtime/CachedBytecode.cpp
 runtime/CachedSpecialPropertyAdaptiveStructureWatchpoint.cpp
-runtime/CachedTypes.cpp
```
You'll need to add this file to `Source/JavaScriptCore/CMakeLists.txt` to keep `cmake` (e.g. linux) builds working.

MenloDorian commented on Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
```diff
@@ -2173,6 +2176,20 @@
+			filePatterns = "*/JSCBytecodeCacheVersion.cpp.in";
+			fileType = pattern.proxy;
+			inputFiles = (
+				"$(OBJECT_FILE_DIR_normal)/$(arch)/CachedTypes.o",
```
Is it possible to add `DerivedSources/JavaScriptCore/JSCBuiltins.cpp` to this list of dependencies?  I think the original https://commits.webkit.org/217978@main that change the cache ID from using build time to `__TIMESTAMP__` actually introduced a bug.  Can you also include the submitted version string in the hash?

MenloDorian reviewed (approved):
r=me

emw-apple committed:
Avoid using __TIMESTAMP__ when building for a compilation cache
https://bugs.webkit.org/show_bug.cgi?id=247622
rdar://102056857

Reviewed by Mark Lam.

For Xcode-based ports, add a build rule to generate a
"JSCBytecodeCacheVersion.cpp" source file, which contains a SHA1 hash of
CachedTypes.o and JSCBuiltins.o, plus a production source version string
if specified. This is a belt-and-suspenders approach to try and ensure
that the version hash changes any time the JSC sources change, while
still being reproducible.

Because of its input and output dependencies, XCBuild runs the build
rule after CachedTypes.cpp and JSCBuiltins.cpp are compiled, then
subsequently compiles the source file the build rule generates.

For CMake + Ninja, it doesn't seem to be easily possible to do achieve
the same build ordering without splitting CachedTypes and JSCBuiltins
out to a separate target. For now, continue to use __TIMESTAMP__.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj: Add
  build rule and source references.
* Source/JavaScriptCore/Sources.txt: Remove CachedTypes.cpp and
  JSCBuiltins.cpp from the unified source set, so that they get their
  own object files.
* Source/JavaScriptCore/runtime/CachedTypes.cpp:
(JSC::GenericCacheEntry::isUpToDate const):
(JSC::jscBytecodeCacheVersion): Deleted.
* Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.cpp.in: Added.
  Template file, contains a CACHED_TYPES_CKSUM token which is replaces
  with the actual checksum by the build rule.
* Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.h: Added.
* Source/ThirdParty/libwebrtc/Source/third_party/yasm/frontends/yasm/yasm.c:
  Remove __DATE__ macro used in assembler output.

Canonical link: https://commits.webkit.org/256816@main
